Unbreak Docker tests

jacobpenny opened the pull request:
Our tests on Travis and the Docker based test suite got out of sync, this brings them to parity. This PR also changes our travis.yml so that it uses Docker.

- Disables sandbox mode (as it is in Travis)
- Added a my.conf in order use the same settings as Travis mysql does (The docker mysql image has stricter defaults). This was causing some tests to fail.
- Update TestUser insert statement (Password_MD5 -> Password_hash)

jacobpenny commented:
We should consider using Docker within Travis.

driusan commented:
What would be involved in using docker within Travis?

And how would it affect the time that a CI run takes?

kongtiaowang commented:
@jacobpenny 
After docker-compose build, I ran "npm run tests:integration".
It shows can't connect.
![may-24-2017 11-10-49](https://cloud.githubusercontent.com/assets/9876007/26410573/b616ace6-4071-11e7-9d22-9507bfb8ff5e.gif)

jacobpenny commented:
@kongtiaowang Hmm, that functionality hasn't changed and still works for me. It is possible that the selenium container wasn't ready when you tried to connect.

jacobpenny commented:
@driusan It would just involve changing the .travis.yml file a bit. I'll try it out and see how it affects the run time.

kongtiaowang commented:
It is very helpful for me.

driusan commented on Dockerfile.test.db:
```diff
@@ -17,9 +17,14 @@ RUN echo "Use LorisTest;" | cat - \
 RUN echo "Use LorisTest;" >> /docker-entrypoint-initdb.d/0001-paths.sql
 RUN echo "UPDATE Config SET Value='${BASE_DIR}/' WHERE ConfigID=(SELECT ID FROM ConfigSettings WHERE Name='base');" >> /docker-entrypoint-initdb.d/0001-paths.sql
 RUN echo "UPDATE Config SET Value='http://web:8000' WHERE ConfigID=(SELECT ID FROM ConfigSettings WHERE Name='url');" >> /docker-entrypoint-initdb.d/0001-paths.sql
+RUN echo "UPDATE users SET Password_hash=CONCAT('aa', MD5('aatestpass')), Pending_approval='N', Password_expiry='2100-01-01' WHERE ID=1;" >> /docker-entrypoint-initdb.d/0002-user.sql
```
This doesn't work (or at least, it shouldn't). Password_hash stores the output of password_hash (which IIRC is bcrypt).

jacobpenny replied:
Hmmm, the tests are working fine. I'll have to look into this.

jacobpenny replied:
Good catch, this line was doing nothing useful. Removed!

driusan commented on test/mysql-config/my.cnf:
```diff
@@ -0,0 +1,2 @@
+[mysqld]
+sql_mode = "NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION"
```
can you put a comment here explaining what this does and why it's needed for the tests?

jacobpenny replied:
Done

driusan replied:
Now that I see why it's done, the `only_full_group_by` is the default in MySQL 5.7 (which is probably why it's happening in Docker). We should probably fix these bugs rather than hiding them.

jacobpenny replied:
I completely agree.

jacobpenny commented:
@driusan Addressed your comments. Also modified `.travis.yml` to use docker. Running times are roughly the same (hard to say exactly, there is a lot of variance).  Have a look at the build here:
https://travis-ci.org/aces/Loris/builds/236527994

Notes:
 - Only tests with PHP 7
 - Breaks up linting/unit/integration into different build jobs

jstirling91 commented:
Removing Passed Manual Tests based on location in project

driusan commented on .travis.yml:
```diff
@@ -1,97 +1,28 @@
-
-    # Run unit tests to make sure functions still do what they should.
-    - vendor/bin/phpunit --configuration test/phpunit.xml
+    - npm run --silent $TEST_SUITE
```
why --silent? It's not like you're sitting there watching it, Travis collapses it when it's finished anyways, and there may be useful information for debugging when something fails.

driusan commented:
Is it possible to do 7.0 and 7.1 with Docker? Dropping 7.1 from automated testing seems problematic, since we'll eventually need to upgrade past 7.0 and it would be better to be proactive.

And is it possible to split this into two PRs? One for the build matrix which splits the different parts of the build into different processes (which I think should be uncontroversial) and one which changes the test suite to docker (which seems to be affecting a lot more than expected)?

kongtiaowang commented:
Please merge this "https://github.com/aces/Loris/pull/2852" first.
It will fix the testing errors in Docker with Mysql 5.7.

driusan commented:
I'm moving this to the Author Needs To Incorporate Changes board, because it's waiting on @jacobpenny investigating if it's possible to have tests run with both PHP 7.0 and 7.1 with Docker under Travis.

jacobpenny commented:
@driusan I'd like to get this merged ASAP so I'm going to issue a separate PR to transition Travis to Docker  (I've reverted `.travis.yml` in this PR). Can you please merge these fixes?

driusan commented on htdocs/GetStatic.php:
```diff
@@ -33,7 +33,13 @@
 $client = new NDB_Client();
-if ($client->initialize("../project/config.xml") == false) {
+
+$configFile = getenv('LORIS_DB_CONFIG');
```
There's more situations than this to take into account.

I was going to suggest moving https://github.com/aces/Loris/blob/17.1-dev/php/libraries/NDB_Config.class.inc#L71 into a helper function, but now that I'm looking at it.. wouldn't this work if you just changed it to `$client->initialize(null)`?

jacobpenny replied:
Ah yes, thank you. I wrongly assumed that one of the default parameters along the way would override my null. I'll update this.
